### PR TITLE
Stop task execution when all output drivers are finished

### PR DESCRIPTION
**Pull request by dain**
Task only needs to wait for output drivers to finish, since the only
required output from a task is through the output buffer which only
output drivers can write to.

### CODE REVIEW

**Comment by findepi** [comment]
> the only
required output from a task is through the output buffer which only
output drivers can write to.

@sopel39 how does this relate to dynamic filtering?

**Comment by sopel39** [comment]
@findepi I don't think it brakes anything. At worst case dynamic tuples won't be send to coordinator.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlTaskExecution.java [hunk 1]
```diff
@@ -258,6 +259,14 @@ private SqlTaskExecution(
         }
     }
 
+    private void stopAllDriverExecution()
+    {
+        taskExecutor.removeTask(taskHandle);
+        for (DriverFactory factory : allRawDriverFactories) {
+            factory.noMoreDrivers();
+        }
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlTaskExecution.java]: I'm trying to reason about the implication of this out-of-bound closure of DriverFactory. That's when I found this issue.

I believe a commit from 2013 (895c62a9a8283eb7944cd2544ecb38878553fd94) introduced a bug that makes DriverFactory closure not guaranteed.

In `DriverSplitRunner.close`, it should decrement the `pendingCreation` counter in `DriverSplitRunnerFactory` if `driver == null`.

**Comment by haozhun** [comment]
I think it would be a better idea to do the same thing to Lifespan (grouped execution). Diverging the logic for Task and Lifespan is likely going to lead to annoyance.

In order to apply this PR to Lifespan, this would be the high-level logic:

* For the output pipeline, if the PipelineExecutionStrategy is
  * ... is UNGROUPED
    * If "no more output driver" and "# current output driver" = 0, kill all drivers in the task
  * ... is GROUPED
    * For each Lifespan (excl TaskWide), if ..., kill all drivers that belong to the Lifespan
    * If ..., kill all drivers in the task

To "kill all drivers in the task" (as implemented in this PR), `TaskExecutor.doRemoveTask` invokes `taskHandle.destroy()` to get a list of split runners, and call `.destory()` on them to destroy the drivers.

To "kill all drivers that belong to the Lifespan", `taskHandle` must provide a way to get a list of split runners that belong a particular Lifespan. This would require changing TaskHandle to manage its 3 internal data structures in a Lifespan-aware fashion.

* `queuedLeafSplits` will need to be a custom data structure that support the following operations: `addLast(k, v)`, `clear()`, `pollFirst()`, `removeAll(k)`. (i.e. a multimap and queue fusion)
* `runningLeafSplits` and `runningIntermediateSplits` will need to be a `HashMultimap`.

If this sounds like a reasonable complexity for this problem, I can add that. Alternatively, we can fix #9357, which would make this PR obsolete.

**Pull request closed without merging**